packages/utreexo/src/stump/proof.cairo

```diff
@@ -27,3 +29,340 @@ impl UtreexoBatchProofDisplay of Display<UtreexoBatchProof> {
         Result::Ok(())
     }
 }
+
+#[generate_trait]
+pub impl UtreexoBatchProofImpl of UtreexoBatchProofTrait {
+    /// Computes a set of roots given a proof and leaves hashes.
+    fn compute_roots(
+        self: @UtreexoBatchProof, del_hashes: Span<felt252>, num_leaves: u64,
+    ) -> Result<Span<felt252>, ByteArray> {
+        let total_rows: u8 = if (num_leaves == 0) {
+            0
+        } else {
+            64 - leading_zeros((num_leaves - 1))
+        };
+        let mut calculated_root_hashes: Array<felt252> = array![];
+        let proof_positions: Span<u64> = get_proof_positions(*self.targets, num_leaves, total_rows);
+
+        let mut nodes: Array<(u64, felt252)> = array![];
+
+        // Append targets with their hashes
+        let mut i = 0;
+        while i != (*self.targets).len() {
+            let pos = *self.targets[i];
+            nodes.append((pos, *del_hashes[i]));
+            i += 1;
+        };
+
+        // Append proof positions with their hashes
+        while i != proof_positions.len() {
+            let pos = *proof_positions[i];
+            nodes.append((pos, *self.proof[i]));
+            i += 1;
+        };
+
+        // Sort nodes for sibling computation
```

Comment:
This is implemented in https://github.com/keep-starknet-strange/raito/pull/253/files#diff-ee5ac2041a9ab94d55b9af69035a3eeac08314326e11ab871821b75b21a13724R85

I think we can sort the nodes as a first step before verifying / deleting to not doing this twice

cc @Groxan